core/adapters/http.go

```diff
@@ -169,54 +170,70 @@ func sendRequest(input models.RunInput, request *http.Request, config HTTPReques
 	}
 	client := &http.Client{Transport: tr}
 
-	response, err := withRetry(client, request, config)
-
+	responseBody, statusCode, err := withRetry(client, request, config)
 	if err != nil {
 		return models.NewRunOutputError(err)
 	}
 
-	defer response.Body.Close()
-
-	source := newMaxBytesReader(response.Body, config.sizeLimit)
-	bytes, err := ioutil.ReadAll(source)
-	if err != nil {
-		return models.NewRunOutputError(err)
-	}
-
-	responseBody := string(bytes)
-	if response.StatusCode >= 400 {
+	// 4xx errors are not retried
+	// We assume this is an error on our side, so retrying will not make any difference
+	if statusCode >= 400 {
 		return models.NewRunOutputError(errors.New(responseBody))
 	}
 
 	return models.NewRunOutputCompleteWithResult(responseBody)
 }
 
+// withRetry executes the http request in a retry. Timeout is controlled with a context
+// Retry occurs if the request timeout, or there is any kind of connection or transport-layer error
+// Retry also occurs on remote server 5XX errors
 func withRetry(
 	client *http.Client,
 	originalRequest *http.Request,
 	config HTTPRequestConfig,
-) (*http.Response, error) {
-	var response *http.Response
-	err := retry.Do(
+) (responseBody string, statusCode int, err error) {
+	err = retry.Do(
 		func() error {
 			ctx, cancel := context.WithTimeout(context.Background(), config.timeout)
 			defer cancel()
 			requestWithTimeout := originalRequest.Clone(ctx)
 
+			start := time.Now()
+
 			r, err := client.Do(requestWithTimeout)
 			if err != nil {
 				return err
 			}
-			response = r
+			defer r.Body.Close()
+			statusCode = r.StatusCode
+			elapsed := time.Since(start)
+			logger.Debugw(fmt.Sprintf("http adapter got %v in %s", statusCode, elapsed), "statusCode", statusCode, "timeElapsedSeconds", elapsed)
+
+			source := newMaxBytesReader(r.Body, config.sizeLimit)
+			bytes, err := ioutil.ReadAll(source)
+			if err != nil {
```

Comment:
This could be problematic if we're using a non-idempotent HTTP verb like `POST` and the connection gets interrupted for any reason after the server already returns a 200 OK. Might be better to not retry in that case and simply mark the run as errored.

Reply:
Given CAP theorem principles it's impossible to POST exactly once. We don't know if the remote server will execute the action or not if the connection is interrupted.

Client may have at most once or at least once. The current default with retry is at least once, I think it's reasonable to stick with that in this case.

Reply:
I feel uneasy with "at-least-once" in this context. Who knows that the request might be doing...

OTOH, as you point out, any client already has to be able to deal with the possibility that a partition occurred during the request and can implement retry logic at a higher level where they are aware of the semantics of the underlying application.

Reply:
Anyways, I'm fine with taking the change under discussion out of scope for the hotfix and moving it into a separate feature we can roll out with a normal release.

Reply:
https://www.pivotaltracker.com/story/show/172202268